text/3617-precise-capturing.md

```diff
@@ -0,0 +1,670 @@
+
+Two, while ATPIT is nearing stabilization, there are yet some type systems details being resolved.  For TAIT, there is much work yet to do.  Putting these features in the critical path would add risk to the edition, to the Lifetime Capture Rules 2024, and to these features.
+
+Three, as a practical matter, an explicit `impl use<..> Trait` syntax lets us write much better automatic migration lints and offers a much more straightforward migration story for our users.
+
+Four, the set of generic parameters that are captured by an opaque type is a fundamental and practical property of that opaque type.  In a language like Rust, it *feels* like there ought to be an explicit syntax for it.  We probably want this in any world.
+
+## Inferred precise capturing
+
+We had hoped that we might be able to achieve something with a similar effect to precise capturing at the cost of an extra generic lifetime parameter in each signature with improvements to the type system.  The goal would be to allow, e.g., this code to work rather than error:
+
+```rust
+fn foo<'o, T>(_: T) -> impl Sized + 'o {}
+
+fn bar(x: ()) -> impl Sized + 'static {
+    foo(&x)
+//~^ ERROR returns a value referencing data owned by the
+//~|       current function
+}
+```
+
+The idea is that, even though the opaque type returned by `foo` does capture the generic type parameter `T`, since the opaque type is explicitly bounded by `'o` and the signature does not assert `T: 'o`, we know that the hidden type cannot actually use `T`.
+
+As it turns out, making full use of this observation is challenging (see [#116040][] and [#116733][]).  While we did make improvements to the type system here, and while more might be possible, this does not solve the problem today in all important cases (including, e.g., avoiding the capture of higher ranked lifetimes in nested opaque types) and will not for the foreseeable future.
+
+Moreover, even with the fullest possible version of these improvements, whether or not a generic parameter is captured by an opaque type would remain observable.  Having an explicit syntax to control what is captured is more direct, more expressive, and leads to a better migration story.
+
+See [Appendix G][] in [RFC 3498][] for more details.
+
+[#116040]: https://github.com/rust-lang/rust/pull/116040
+[#116733]: https://github.com/rust-lang/rust/pull/116733
+[Appendix G]: https://github.com/rust-lang/rfcs/blob/master/text/3498-lifetime-capture-rules-2024.md#appendix-g-future-possibility-inferred-precise-capturing
+
+## Syntax
+
+We considered a number of different possible syntaxes before landing on `impl use<..> Trait`.  We'll discuss each considered.
+
+### `impl use<..> Trait`
```

Comment:
I suspect that it's going to get annoying pretty quick (compared to `impl<..> Trait`), if precise captures become something that is specified often (e.g. by convention to minimize captures).